units/data_cleanup_testresults.go

```diff
@@ -0,0 +1,140 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/testresult"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/mongodb/grip/sometimes"
+)
+
+const testResultsCleanupJobName = "data-cleanup-testresults"
+
+func init() {
+	registry.AddJobType(testResultsCleanupJobName, func() amboy.Job {
+		return makeTestResultsCleanupJob()
+	})
+}
+
+type dataCleanupTestResults struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	env evergreen.Environment
+}
+
+func makeTestResultsCleanupJob() *dataCleanupTestResults {
+	j := &dataCleanupTestResults{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    testResultsCleanupJobName,
+				Version: 0,
+			},
+		},
+	}
+
+	j.SetDependency(dependency.NewAlways())
+
+	return j
+}
+
+func NewTestResultsCleanupJob(ts time.Time) amboy.Job {
+	j := makeTestResultsCleanupJob()
+	j.SetID(fmt.Sprintf("%s.%s", testResultsCleanupJobName, ts.Format(TSFormat)))
+	return j
+}
+
+func (j *dataCleanupTestResults) Run(ctx context.Context) {
+	defer j.MarkComplete()
+	startAt := time.Now()
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	flags, err := evergreen.GetServiceFlags()
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+
+	if flags.BackgroundCleanup {
+		grip.InfoWhen(sometimes.Percent(evergreen.DegradedLoggingPercent), message.Fields{
+			"job_type": testResultsCleanupJobName,
+			"job_id":   j.ID(),
+			"message":  "disaster recovery backups disabled, also disabling cleanup",
+		})
+		return
+	}
+
+	var (
+		offsets   []int
+		batches   int
+		numDocs   int
+		timeSpent time.Duration
+	)
+
+	totalDocs, _ := j.env.DB().Collection(testresult.Collection).EstimatedDocumentCount(ctx)
+
+	lock := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	for i := 0; i < 10; i++ {
```

Comment:
I tested this with 100k at a time, but depending on what the bottleneck is, this could take 10x the time with 10 threads.

Reply:
it runs as many threads as it can for 45ish seconds, cools down, and then another job runs.

Reply:
Well, it runs 10 threads, not "as it can". And it runs all of them no matter what. If one takes 20 seconds 10 could take much longer than 45 for a single iteration.

Reply:
we could put a scope on it?

Reply:
the worst thing that happens is that it takes a minute, everything dies after 45

Reply:
I'm concerned that these 10 threads take 200 seconds, so jobs back up. Adding a scope will not prevent jobs from backing up.

Reply:
no single thread can take more than 44s+(1deleteWithLimitCall), and the entire job can take no longer than the thread.